File: elephanto-events/backend/models/velvet_hour.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// VelvetHourSession represents a Velvet Hour session
type VelvetHourSession struct {
	ID             uuid.UUID  `json:"id" db:"id"`
	EventID        uuid.UUID  `json:"eventId" db:"event_id"`
	StartedAt      time.Time  `json:"startedAt" db:"started_at"`
	EndedAt        *time.Time `json:"endedAt" db:"ended_at"`
	IsActive       bool       `json:"isActive" db:"is_active"`
	CurrentRound   int        `json:"currentRound" db:"current_round"`
	RoundStartedAt *time.Time `json:"roundStartedAt" db:"round_started_at"`
	RoundEndsAt    *time.Time `json:"roundEndsAt" db:"round_ends_at"`
	Status         string     `json:"status" db:"status"` // waiting, in_round, break, completed
	CreatedAt      time.Time  `json:"createdAt" db:"created_at"`
	UpdatedAt      time.Time  `json:"updatedAt" db:"updated_at"`
}

// VelvetHourParticipant represents a user participating in Velvet Hour
type VelvetHourParticipant struct {
	ID        uuid.UUID `json:"id" db:"id"`
	SessionID uuid.UUID `json:"sessionId" db:"session_id"`
	UserID    uuid.UUID `json:"userId" db:"user_id"`
	JoinedAt  time.Time `json:"joinedAt" db:"joined_at"`
	Status    string    `json:"status" db:"status"` // waiting, matched, in_round, completed
	CreatedAt time.Time `json:"createdAt" db:"created_at"`
	UpdatedAt time.Time `json:"updatedAt" db:"updated_at"`
	
	// Joined user information
	UserName  string `json:"userName" db:"user_name"`
	UserEmail string `json:"userEmail" db:"user_email"`
}

// VelvetHourMatch represents a match between two users
type VelvetHourMatch struct {
	ID            uuid.UUID  `json:"id" db:"id"`
	SessionID     uuid.UUID  `json:"sessionId" db:"session_id"`
	RoundNumber   int        `json:"roundNumber" db:"round_number"`
	User1ID       uuid.UUID  `json:"user1Id" db:"user1_id"`
	User2ID       uuid.UUID  `json:"user2Id" db:"user2_id"`
	MatchNumber   int        `json:"matchNumber" db:"match_number"`
	MatchColor    string     `json:"matchColor" db:"match_color"`
	StartedAt     *time.Time `json:"startedAt" db:"started_at"`
	ConfirmedUser1 bool      `json:"confirmedUser1" db:"confirmed_user1"`
	ConfirmedUser2 bool      `json:"confirmedUser2" db:"confirmed_user2"`
	ConfirmedAt   *time.Time `json:"confirmedAt" db:"confirmed_at"`
	CreatedAt     time.Time  `json:"createdAt" db:"created_at"`
	UpdatedAt     time.Time  `json:"updatedAt" db:"updated_at"`
	
	// Joined user information
	User1Name string `json:"user1Name" db:"user1_name"`
	User2Name string `json:"user2Name" db:"user2_name"`
}

// VelvetHourFeedback represents feedback from one user about another
type VelvetHourFeedback struct {
	ID             uuid.UUID `json:"id" db:"id"`
	MatchID        uuid.UUID `json:"matchId" db:"match_id"`
	FromUserID     uuid.UUID `json:"fromUserId" db:"from_user_id"`
	ToUserID       uuid.UUID `json:"toUserId" db:"to_user_id"`
	WantToConnect  bool      `json:"wantToConnect" db:"want_to_connect"`
	FeedbackReason string    `json:"feedbackReason" db:"feedback_reason"`
	SubmittedAt    time.Time `json:"submittedAt" db:"submitted_at"`
	CreatedAt      time.Time `json:"createdAt" db:"created_at"`
}

// VelvetHourQuestion represents configurable questions for Velvet Hour
type VelvetHourQuestion struct {
	ID           uuid.UUID `json:"id" db:"id"`
	EventID      uuid.UUID `json:"eventId" db:"event_id"`
	QuestionType string    `json:"questionType" db:"question_type"`
	QuestionText string    `json:"questionText" db:"question_text"`
	Options      string    `json:"options" db:"options"` // JSON string for options
	DisplayOrder int       `json:"displayOrder" db:"display_order"`
	CreatedAt    time.Time `json:"createdAt" db:"created_at"`
	UpdatedAt    time.Time `json:"updated_at" db:"updated_at"`
}

// Request/Response models for Velvet Hour

type CreateVelvetHourSessionRequest struct {
	EventID uuid.UUID `json:"eventId" validate:"required"`
}

type VelvetHourSessionResponse struct {
	Session      *VelvetHourSession      `json:"session"`
	Participants []VelvetHourParticipant `json:"participants"`
	CurrentMatch *VelvetHourMatch        `json:"currentMatch,omitempty"`
}

type JoinVelvetHourRequest struct {
	// No additional fields needed - user comes from auth context
}

type VelvetHourStatusResponse struct {
	IsActive     bool                   `json:"isActive"`
	Session      *VelvetHourSession     `json:"session,omitempty"`
	Participant  *VelvetHourParticipant `json:"participant,omitempty"`
	CurrentMatch *VelvetHourMatch       `json:"currentMatch,omitempty"`
	TimeLeft     *int                   `json:"timeLeft,omitempty"` // seconds remaining
}

type ConfirmMatchRequest struct {
	MatchID uuid.UUID `json:"matchId" validate:"required"`
}

type SubmitFeedbackRequest struct {
	MatchID        uuid.UUID `json:"matchId" validate:"required"`
	WantToConnect  bool      `json:"wantToConnect"`
	FeedbackReason string    `json:"feedbackReason" validate:"required"`
}

type AdminVelvetHourStatusResponse struct {
	Session         *VelvetHourSession      `json:"session"`
	Participants    []VelvetHourParticipant `json:"participants"`
	CurrentMatches  []VelvetHourMatch       `json:"currentMatches"`
	CompletedRounds int                     `json:"completedRounds"`
	CanStartRound   bool                    `json:"canStartRound"`
}

type StartRoundRequest struct {
	Matches []ManualMatch `json:"matches,omitempty"` // Optional manual matches
}

type ManualMatch struct {
	User1ID     uuid.UUID `json:"user1Id"`
	User2ID     uuid.UUID `json:"user2Id"`
	MatchNumber int       `json:"matchNumber"`
	MatchColor  string    `json:"matchColor"`
}

type UpdateVelvetHourConfigRequest struct {
	RoundDuration     *int `json:"roundDuration"`
	BreakDuration     *int `json:"breakDuration"`
	TotalRounds       *int `json:"totalRounds"`
	MinParticipants   *int `json:"minParticipants"`
}

File: elephanto-events/frontend/src/types/velvet-hour.ts
// Velvet Hour types

export interface VelvetHourSession {
  id: string;
  eventId: string;
  startedAt: string;
  endedAt?: string;
  isActive: boolean;
  currentRound: number;
  roundStartedAt?: string;
  roundEndsAt?: string;
  status: string; // waiting, in_round, break, completed
  createdAt: string;
  updatedAt: string;
}

export interface VelvetHourParticipant {
  id: string;
  sessionId: string;
  userId: string;
  joinedAt: string;
  status: string; // waiting, matched, in_round, completed
  createdAt: string;
  updatedAt: string;
  userName: string;
  userEmail: string;
}

export interface VelvetHourMatch {
  id: string;
  sessionId: string;
  roundNumber: number;
  user1Id: string;
  user2Id: string;
  matchNumber: number;
  matchColor: string;
  startedAt?: string;
  confirmedUser1: boolean;
  confirmedUser2: boolean;
  confirmedAt?: string;
  createdAt: string;
  updatedAt: string;
  user1Name: string;
  user2Name: string;
  user1FeedbackSubmitted?: boolean;
  user2FeedbackSubmitted?: boolean;
}

export interface VelvetHourFeedback {
  id: string;
  matchId: string;
  fromUserId: string;
  toUserId: string;
  wantToConnect: boolean;
  feedbackReason: string;
  submittedAt: string;
  createdAt: string;
}

// API Request/Response types

export interface VelvetHourStatusResponse {
  isActive: boolean;
  session?: VelvetHourSession;
  participant?: VelvetHourParticipant;
  currentMatch?: VelvetHourMatch;
  timeLeft?: number; // seconds remaining
  config?: VelvetHourConfig;
}

export interface SubmitFeedbackRequest {
  matchId: string;
  wantToConnect: boolean;
  feedbackReason: string; // humor, confidence, listening, no_connection
}

export interface VelvetHourConfig {
  roundDuration: number;
  breakDuration: number;
  totalRounds: number;
  minParticipants: number;
}

export interface AdminVelvetHourStatusResponse {
  session?: VelvetHourSession;
  participants: VelvetHourParticipant[];
  currentMatches: VelvetHourMatch[];
  completedRounds: number;
  canStartRound: boolean;
  config: VelvetHourConfig;
}

export interface StartRoundRequest {
  matches?: ManualMatch[];
}

export interface ManualMatch {
  user1Id: string;
  user2Id: string;
  matchNumber: number;
  matchColor: string;
}

export interface UpdateVelvetHourConfigRequest {
  roundDuration?: number;
  breakDuration?: number;
  totalRounds?: number;
  // minParticipants is auto-calculated based on totalRounds
}

// Frontend component props types

export interface VelvetHourWaitingProps {
  onJoin: () => void;
  participantCount: number;
  hasJoined: boolean;
  isConnected: boolean;
  config?: VelvetHourConfig;
  isInActiveRound?: boolean;
  currentRound?: number;
  timeLeft?: number;
  partnerName?: string;
  matchNumber?: number;
  matchColor?: string;
}

export interface VelvetHourMatchProps {
  match: VelvetHourMatch;
  onConfirmMatch: (matchId: string) => void;
  currentUserId: string;
}

export interface VelvetHourRoundProps {
  match: VelvetHourMatch;
  timeLeft: number;
  currentRound: number;
  totalRounds: number;
  currentUserId: string;
  roundDuration?: number; // Optional, defaults to 10 minutes
}

export interface VelvetHourFeedbackProps {
  match: VelvetHourMatch;
  onSubmitFeedback: (matchId: string, wantToConnect: boolean, reason: string) => void;
  currentUserId: string;
}

export interface VelvetHourCompleteProps {
  // Add props if needed for showing connections/results
}

// Admin component props

export interface AdminVelvetHourControlProps {
  eventId: string;
  eventTitle: string;
  eventDate: string;
  eventTime: string;
  status: AdminVelvetHourStatusResponse;
  onStartSession: () => void;
  onStartRound: (matches?: ManualMatch[]) => void;
  onEndSession: () => void;
  onUpdateConfig: (config: UpdateVelvetHourConfigRequest) => void;
  onResetSession?: () => void;
}

export interface DraggableMatchmakingProps {
  participants: VelvetHourParticipant[];
  onMatchesChange: (matches: ManualMatch[]) => void;
  onMatchStatsChange?: (stats: { incompleteMatches: number; completeMatches: number; duplicatePairings: number; validationErrors: string[] }) => void;
  maxMatches: number;
  eventId: string;
  previousMatches?: VelvetHourMatch[]; // All previous matches to check for duplicates
}

// Color options for matches (supports up to 100 rooms)
export const MATCH_COLORS = [
  'red', 'blue', 'green', 'purple', 'orange', 'yellow', 'pink', 'cyan',
  'indigo', 'teal', 'lime', 'rose', 'amber', 'emerald', 'violet', 'sky',
  'slate', 'gray', 'zinc', 'neutral', 'stone', 'red-dark', 'rose-dark', 'pink-dark',
  'fuchsia', 'purple-dark', 'violet-dark', 'indigo-dark', 'blue-dark', 'sky-dark',
  'cyan-dark', 'teal-dark', 'emerald-dark', 'green-dark', 'lime-dark', 'yellow-dark',
  'amber-dark', 'orange-dark', 'red-light', 'rose-light', 'pink-light', 'fuchsia-light',
  'purple-light', 'violet-light', 'indigo-light', 'blue-light', 'sky-light', 'cyan-light',
  'teal-light', 'emerald-light', 'green-light', 'lime-light', 'yellow-light', 'amber-light',
  'orange-light', 'red-bright', 'rose-bright', 'pink-bright', 'fuchsia-bright', 'purple-bright',
  'violet-bright', 'indigo-bright', 'blue-bright', 'sky-bright', 'cyan-bright', 'teal-bright',
  'emerald-bright', 'green-bright', 'lime-bright', 'yellow-bright', 'amber-bright', 'orange-bright',
  'red-deep', 'rose-deep', 'pink-deep', 'fuchsia-deep', 'purple-deep', 'violet-deep',
  'indigo-deep', 'blue-deep', 'sky-deep', 'cyan-deep', 'teal-deep', 'emerald-deep',
  'green-deep', 'lime-deep', 'yellow-deep', 'amber-deep', 'orange-deep', 'brown',
  'tan', 'beige', 'cream', 'ivory', 'pearl', 'silver', 'gold', 'bronze',
  'copper', 'rust', 'maroon', 'crimson', 'scarlet', 'burgundy', 'wine', 'navy'
] as const;

export type MatchColor = typeof MATCH_COLORS[number];

// Feedback reasons
export const FEEDBACK_REASONS = [
  { value: 'humor', label: 'Their sense of humor' },
  { value: 'confidence', label: 'Their confidence and presence' },
  { value: 'listening', label: 'The way they listened and engaged' },
  { value: 'no_connection', label: "We didn't really connect" }
] as const;